File: libs/document-processor-next/src/runtime/gas-meter.ts
import type { Node } from '../types/index.js';
import { canonicalSize } from '../util/node-canonicalizer.js';
import { normalizeScope } from '../util/pointer-utils.js';

const INITIALIZATION = 1_000;
const CHANNEL_MATCH_ATTEMPT = 5;
const HANDLER_OVERHEAD = 50;
const BOUNDARY_CHECK = 2;
const PATCH_REMOVE = 10;
const BRIDGE_NODE = 10;
const DRAIN_EVENT = 10;
const CHECKPOINT_UPDATE = 20;
const TERMINATION_MARKER = 20;
const LIFECYCLE_DELIVERY = 30;
const FATAL_TERMINATION_OVERHEAD = 100;

function scopeEntryCharge(depth: number): number {
  return 50 + 10 * depth;
}

function patchAddOrReplaceCharge(sizeCharge: number): number {
  return 20 + sizeCharge;
}

function cascadeRoutingCharge(scopeCount: number): number {
  return 10 * scopeCount;
}

function emitEventCharge(sizeCharge: number): number {
  return 20 + sizeCharge;
}

export class GasMeter {
  private total = 0;

  totalGas(): number {
    return this.total;
  }

  add(amount: number): void {
    this.total += amount;
  }

  chargeScopeEntry(scopePath: string): void {
    this.add(scopeEntryCharge(this.scopeDepth(scopePath)));
  }

  chargeInitialization(): void {
    this.add(INITIALIZATION);
  }

  chargeChannelMatchAttempt(): void {
    this.add(CHANNEL_MATCH_ATTEMPT);
  }

  chargeHandlerOverhead(): void {
    this.add(HANDLER_OVERHEAD);
  }

  chargeBoundaryCheck(): void {
    this.add(BOUNDARY_CHECK);
  }

  chargePatchAddOrReplace(value: Node | null | undefined): void {
    this.add(patchAddOrReplaceCharge(this.payloadSizeCharge(value)));
  }

  chargePatchRemove(): void {
    this.add(PATCH_REMOVE);
  }

  chargeCascadeRouting(scopeCount: number): void {
    if (scopeCount > 0) {
      this.add(cascadeRoutingCharge(scopeCount));
    }
  }

  chargeEmitEvent(event: Node | null | undefined): void {
    this.add(emitEventCharge(this.payloadSizeCharge(event)));
  }

  chargeBridge(): void {
    this.add(BRIDGE_NODE);
  }

  chargeDrainEvent(): void {
    this.add(DRAIN_EVENT);
  }

  chargeCheckpointUpdate(): void {
    this.add(CHECKPOINT_UPDATE);
  }

  chargeTerminationMarker(): void {
    this.add(TERMINATION_MARKER);
  }

  chargeLifecycleDelivery(): void {
    this.add(LIFECYCLE_DELIVERY);
  }

  chargeFatalTerminationOverhead(): void {
    this.add(FATAL_TERMINATION_OVERHEAD);
  }

  private payloadSizeCharge(node: Node | null | undefined): number {
    if (!node) {
      return 0;
    }
    const bytes = canonicalSize(node);
    return Math.floor((bytes + 99) / 100);
  }

  private scopeDepth(scopePath: string): number {
    const normalized = normalizeScope(scopePath);
    if (normalized === '/' || normalized.length <= 1) {
      return 0;
    }
    let depth = 1;
    for (let i = 1; i < normalized.length; i += 1) {
      if (normalized.charAt(i) === '/') {
        depth += 1;
      }
    }
    return depth;
  }
}

File: references/java-sdk/src/main/java/blue/language/processor/GasMeter.java
package blue.language.processor;

import blue.language.model.Node;
import blue.language.processor.util.NodeCanonicalizer;
import blue.language.processor.util.PointerUtils;

import java.math.BigInteger;

/**
 * Tracks and charges gas usage for a processing run.
 */
final class GasMeter {

    private long totalGas;

    long totalGas() {
        return totalGas;
    }

    void add(long amount) {
        totalGas += amount;
    }

    void chargeScopeEntry(String scopePath) {
        add(GasCharges.scopeEntry(PointerUtils.splitPointerSegments(scopePath).length));
    }

    void chargeInitialization() {
        add(GasCharges.INITIALIZATION);
    }

    void chargeChannelMatchAttempt() {
        add(GasCharges.CHANNEL_MATCH_ATTEMPT);
    }

    void chargeHandlerOverhead() {
        add(GasCharges.HANDLER_OVERHEAD);
    }

    void chargeBoundaryCheck() {
        add(GasCharges.BOUNDARY_CHECK);
    }

    void chargePatchAddOrReplace(Node value) {
        add(GasCharges.patchAddOrReplace(payloadSizeCharge(value)));
    }

    void chargePatchRemove() {
        add(GasCharges.PATCH_REMOVE);
    }

    void chargeCascadeRouting(int scopeCount) {
        if (scopeCount > 0) {
            add(GasCharges.cascadeRouting(scopeCount));
        }
    }

    void chargeEmitEvent(Node event) {
        add(GasCharges.emitEvent(payloadSizeCharge(event)));
    }

    void chargeBridge(Node event) {
        add(GasCharges.BRIDGE_NODE);
    }

    void chargeDrainEvent() {
        add(GasCharges.DRAIN_EVENT);
    }

    void chargeCheckpointUpdate() {
        add(GasCharges.CHECKPOINT_UPDATE);
    }

    void chargeTerminationMarker() {
        add(GasCharges.TERMINATION_MARKER);
    }

    void chargeLifecycleDelivery() {
        add(GasCharges.LIFECYCLE_DELIVERY);
    }

    void chargeFatalTerminationOverhead() {
        add(GasCharges.FATAL_TERMINATION_OVERHEAD);
    }

    void chargeTriggerEventBase() {
        add(GasCharges.TRIGGER_EVENT_BASE);
    }

    void chargeUpdateDocumentBase(int changesetLength) {
        add(GasCharges.updateDocumentBase(changesetLength));
    }

    void chargeDocumentSnapshot(String absolutePointer, Node snapshot) {
        long bytes = snapshot != null ? NodeCanonicalizer.canonicalSize(snapshot) : 0L;
        add(GasCharges.documentSnapshot(absolutePointer, bytes));
    }

    void chargeWasmGas(BigInteger wasmFuel) {
        if (wasmFuel == null) {
            return;
        }
        long charge = ProcessorGasSchedule.wasmFuelToHostGas(wasmFuel);
        if (charge > 0L) {
            add(charge);
        }
    }

    void chargeWasmGas(long wasmFuel) {
        if (wasmFuel <= 0L) {
            return;
        }
        chargeWasmGas(BigInteger.valueOf(wasmFuel));
    }

    private long payloadSizeCharge(Node node) {
        long bytes = NodeCanonicalizer.canonicalSize(node);
        return GasCharges.ceil100(bytes);
    }

    private static final class GasCharges {
        private static final long INITIALIZATION = 1_000L;
        private static final long CHANNEL_MATCH_ATTEMPT = 5L;
        private static final long HANDLER_OVERHEAD = 50L;
        private static final long BOUNDARY_CHECK = 2L;
        private static final long PATCH_REMOVE = 10L;
        private static final long BRIDGE_NODE = 10L;
        private static final long DRAIN_EVENT = 10L;
        private static final long CHECKPOINT_UPDATE = 20L;
        private static final long TERMINATION_MARKER = 20L;
        private static final long LIFECYCLE_DELIVERY = 30L;
        private static final long TRIGGER_EVENT_BASE = 30L;
        private static final long FATAL_TERMINATION_OVERHEAD = 100L;

        private static long scopeEntry(int depth) {
            return 50L + 10L * depth;
        }

        private static long patchAddOrReplace(long sizeCharge) {
            return 20L + sizeCharge;
        }

        private static long cascadeRouting(int scopeCount) {
            return 10L * scopeCount;
        }

        private static long emitEvent(long sizeCharge) {
            return 20L + sizeCharge;
        }

        private static long ceil100(long amount) {
            return (amount + 99L) / 100L;
        }

        private static int pointerDepth(String absolutePointer) {
            String normalized = PointerUtils.normalizePointer(absolutePointer);
            if ("/".equals(normalized)) {
                return 0;
            }
            int depth = 0;
            for (int i = 0; i < normalized.length(); i++) {
                if (normalized.charAt(i) == '/') {
                    depth++;
                }
            }
            return depth;
        }

        private static long updateDocumentBase(int changesetLength) {
            int length = Math.max(0, changesetLength);
            return 40L + 5L * length;
        }

        private static long documentSnapshot(String absolutePointer, long snapshotBytes) {
            return 8L + pointerDepth(absolutePointer) + ceil100(snapshotBytes);
        }
    }
}
